graphql_compiler/compiler/ir_lowering_cypher/ir_lowering.py

```diff
@@ -0,0 +1,185 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+from functools import partial
+
+from ..blocks import CoerceType, Filter, Fold, MarkLocation, Recurse, Traverse
+from ..expressions import (
+    BinaryComposition, ContextField, FoldedContextField, LocalField, NullLiteral
+)
+from ..helpers import get_only_element_from_collection
+from ..ir_lowering_common.common import merge_consecutive_filter_clauses
+from ..ir_lowering_common.location_renaming import (
+    make_location_rewriter_visitor_fn, make_revisit_location_translations
+)
+from ..helpers import FoldScopeLocation
+
+
+##################################
+# Optimization / lowering passes #
+##################################
+
+def insert_explicit_type_bounds(ir_blocks, query_metadata_table, type_equivalence_hints=None):
```

Comment:
In retrospect, doesn't it make more sense to have the type bounds for all vertices in the IR, instead of using `QueryRoot` and `CoerceType`? It seems like all backends use this information differently. For example, the Match backend derives `type_bounds` information from `QueryRoot` and `CoerceType` blocks, and replaces them with `CoerceType` and `Filter` blocks to enforce execution order.

Reply:
If we had a chance to redesign the IR from scratch right now, that would probably have been a good idea. However, I don't think the cost-benefit tradeoff is worth it in this case, since the type of each location is captured in the query metadata already.